### PR TITLE
Move kcfg.minimize into a separate file and class

### PR DESCRIPTION
**Pull request by Stevengre**
Because there will be more pattern rewriting to minimize the KCFG, we need to extract syntax-unrelated functions from `kcfg.py` for clarity. This PR doesn’t introduce any new features; it simply moves functions related to KCFG minimization to `minimize.py`.

The downstream repo that would be affected is likely https://github.com/runtimeverification/kontrol:

- `Proof.minimize_kcfg()` == `KCFGExploration.minimize_kcfg()` will not be affected:
  - https://github.com/runtimeverification/kontrol/blob/beacbb2d8267e29804864364e7263846283e088f/src/kontrol/foundry.py#L988-L992
  - https://github.com/runtimeverification/kontrol/blob/beacbb2d8267e29804864364e7263846283e088f/src/kontrol/prove.py#L463-L464

- `kcfg.minimize` will not be affected:
  - https://github.com/runtimeverification/kontrol/blob/beacbb2d8267e29804864364e7263846283e088f/src/kontrol/prove.py#L730-L734

### CODE REVIEW

File: pyk/src/pyk/kcfg/kcfg.py [hunk 1]
```diff
@@ -1308,6 +1132,9 @@ def write_cfg_data(self) -> None:
         self._deleted_nodes.clear()
         self._created_nodes.clear()
 
+    def minimize(self) -> None:
```
Comment (Stevengre) [inline pyk/src/pyk/kcfg/kcfg.py:1135]: kcfg.minimize() with the same effect.

File: pyk/src/pyk/kcfg/exploration.py [hunk 2]
```diff
@@ -109,4 +111,4 @@ def to_dict(self) -> dict[str, Any]:
 
     # Minimizing the KCFG
     def minimize_kcfg(self) -> None:
-        self.kcfg.minimize()
+        KCFGMinimizer(self.kcfg).minimize()
```
Comment (Stevengre) [inline pyk/src/pyk/kcfg/exploration.py]: proof.minimize_kcfg() with the same effect.

File: pyk/src/pyk/kcfg/kcfg.py [hunk 3]
```diff
@@ -1308,6 +1132,9 @@ def write_cfg_data(self) -> None:
...
+        KCFGMinimizer(self).minimize()
```
Comment (Stevengre) [inline pyk/src/pyk/kcfg/kcfg.py:1136]: kcfg.minimize() with the same effect.

File: pyk/src/pyk/kcfg/exploration.py (hunk 2 above)
Comment (tothtamas28) [reply pyk/src/pyk/kcfg/exploration.py]: This change doesn't seem necessary, as `KCFG` still provides the `minimize` method.

File: pyk/src/pyk/kcfg/minimize.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+from pyk.cterm import CTerm
+from pyk.utils import not_none, single
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+    from pyk.kcfg.kcfg import KCFG, NodeIdLike
+
+
+class KCFGMinimizer:
```
Comment (tothtamas28) [inline pyk/src/pyk/kcfg/minimize.py:15]: Just to confirm: does `KCFG` have any private attributes this class needs access to?

File: pyk/src/tests/integration/proof/test_refute_node.py [hunk 5]
```diff
@@ -372,7 +373,7 @@ def test_apr_proof_refute_node_multiple_constraints(
         prover.advance_proof(proof, max_iterations=4)
         # After the minimization, nodes 7-10 created by the advancement of the proof
         # will have multiple constraints in their immediately preceding splits
-        proof.kcfg.minimize()
+        KCFGMinimizer(proof.kcfg).minimize()
```
Comment (tothtamas28) [inline pyk/src/tests/integration/proof/test_refute_node.py]: Here as well. Let's keep `KCFGMinimizer` private to the `kcfg` module.

File: pyk/src/pyk/kcfg/exploration.py (hunk 2 above)
Comment (Stevengre) [reply pyk/src/pyk/kcfg/exploration.py]: Thank you for pointing this out. I'll remove this change.

File: pyk/src/pyk/kcfg/minimize.py (hunk 4 above)
Comment (Stevengre) [reply pyk/src/pyk/kcfg/minimize.py:15]: self.kcfg._resolve(b_id) at L134

Comment (Stevengre) [reply pyk/src/pyk/kcfg/minimize.py:15]: Does it matter?

Comment (tothtamas28) [reply pyk/src/pyk/kcfg/minimize.py:15]: > Does it matter?

In general, yes.
* There's no guarantee for backwards compatibiliy on a private method.
* A class can only guarantee its invariants if it's internals are not tinkered with.

However, this particular case is pretty innocuous. Still, it'd great to eliminate it (e.g. by `node(b_id).id`).

File: pyk/src/tests/integration/proof/test_refute_node.py (hunk 5 above)
Comment (Stevengre) [reply pyk/src/tests/integration/proof/test_refute_node.py]: Sure, I'll change that!

File: pyk/src/pyk/kcfg/minimize.py (hunk 4 above)
Comment (Stevengre) [reply pyk/src/pyk/kcfg/minimize.py:15]: Thank you for your suggestion! I’ll make that change. But I have a question: will `node(b_id).id` be significantly slower than the current implementation?

File: pyk/src/pyk/kcfg/minimize.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        splits_from_a, splits_from_b = split_from_a.splits, split_from_b.splits
+        a = split_from_a.source
+        ci = list(splits_from_b.keys())
+        # Ensure split can be lifted soundly (i.e., that it does not introduce fresh variables)
+        assert (
+            len(split_from_b.source_vars.difference(a.free_vars)) == 0
+            and len(split_from_b.target_vars.difference(split_from_b.source_vars)) == 0
+        )
+        # Get the substitution for `B`, at the same time removing 'B' from the targets of `A`.
+        csubst_b = splits_from_a.pop(self.kcfg.node(b_id).id)
```
Comment (Stevengre) [inline pyk/src/pyk/kcfg/minimize.py:134]: Done!

File: pyk/src/tests/integration/proof/test_refute_node.py (hunk 5 above)
Comment (Stevengre) [reply pyk/src/tests/integration/proof/test_refute_node.py]: Done! Nothing changed in [test_refute_node.py](https://github.com/runtimeverification/k/pull/4607/files#diff-c50f6d907465456da017ef879b6a11d15c1cfddc76dfc8eb3b6f122532239616) for now.

File: pyk/src/pyk/kcfg/minimize.py (hunk 4 above)
Comment (tothtamas28) [reply pyk/src/pyk/kcfg/minimize.py:15]: It's not significantly slower, it's just one extra map lookup plus an attribute access.

**Review by tothtamas28 — APPROVED** [review]
Nice refactoring!

**Review by anvacaru — APPROVED** [review]

**Pull request merged**